src/main/scala/com/mwz/sonar/scala/scoverage/ScoverageSensor.scala

```diff
@@ -0,0 +1,171 @@
+/*
+ * Sonar Scala Plugin
+ * Copyright (C) 2018 All contributors
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package com.mwz.sonar.scala.scoverage
+
+import com.mwz.sonar.scala.Scala
+import com.mwz.sonar.scala.util.JavaOptionals._
+import com.mwz.sonar.scala.util.PathUtils
+import java.nio.file.Paths
+import org.sonar.api.batch.fs.{FileSystem, InputComponent, InputFile}
+import org.sonar.api.batch.sensor.{Sensor, SensorContext, SensorDescriptor}
+import org.sonar.api.config.Configuration;
+import org.sonar.api.utils.log.Loggers
+import scala.collection.JavaConverters._
+import scala.util.{Failure, Success, Try}
+
+/** Main sensor for importing Scoverage reports to SonarQube */
+final class ScoverageSensor extends ScoverageSensorInternal with ScoverageReportParser
+
+/** Implementation of the sensor */
+private[scoverage] abstract class ScoverageSensorInternal extends Sensor {
+  // cake pattern to mock the scoverage report parser in tests
+  scoverageReportParser: ScoverageReportParserAPI =>
+
+  private[this] val logger = Loggers.get(classOf[ScoverageSensorInternal])
+
+  /** Populates the [[SensorDescriptor]] of this sensor. */
+  override def describe(descriptor: SensorDescriptor): Unit = {
+    descriptor
+      .onlyOnLanguage(Scala.Key)
+      .onlyOnFileType(InputFile.Type.MAIN)
+      .name(ScoverageSensorInternal.SensorName)
+  }
+
+  /** Saves in SonarQube the scoverage information of a module */
+  override def execute(context: SensorContext): Unit = {
```

Comment:
Previously the sensor would aggregate over all the existing modules and create the total coverage for the whole project [here](https://github.com/mwz/sonar-scala/blob/3f025ee80785b5da58afc74a4ba12362a6f6c7af/src/main/scala/com/buransky/plugins/scoverage/sensor/ScoverageSensor.scala#L109), but given that this sensor now only has in the scope one module at a time does this still work as expected? Does Sonar do the coverage aggregation automatically? I'm not entirely sure how this works now for multi-module projects 🤔

Reply:
The `coverage` metric, that is computed internally from sonar  using the  the [_lineHits_](http://javadocs.sonarsource.org/6.7.2/apidocs/org/sonar/api/batch/sensor/coverage/NewCoverage.html#lineHits-int-int--) method in the [**NewCoverage**](http://javadocs.sonarsource.org/6.7.2/apidocs/org/sonar/api/batch/sensor/coverage/NewCoverage.html) class, gets aggregated as expected.
But custom metrics like `scoverage`, and `branch_scoverage` don't.
You can see them by each module.

An alternative, is to make the sensor global, so sonar will execute it for the whole project, but that would mean, that the user should had generate a global scoverage report, or that the sensor should need to open all module's reports and merge the results, like the previous one.

I don't know what you think about it.

Reply:
I run analysis on a sample multi-module project and It seems to me that your approach has exactly the same outcome as the previous approach - I was able to see the same measurements that I get using the existing version of the plugin, so I'm onboard with this way of doing it.